### PR TITLE
[Contracts] Add Cache contract to extend PSR-6 with tag invalidation, callback-based computation and stampede protection

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Let's separate the useful interfaces of the Cache component in Contracts.
This does not include `PruneableInterface`, `MarshallerInterface` nor `*AdapterInterface` because they are too specific to the component.

But `CacheInterface`, `TawAwareCacheInterface` and `ItemInterface` form a nice consistent set of features on top of PSR-6.

Updating the Cache component to use these interfaces will be done in a separate PR so that we can focus on the contract itself here.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Now with a `GetForCacheItemPoolTrait` to help implementing the `CacheInterface::get()` method in PSR-6 implementations.

**Comment by nicolas-grekas** [comment]
Also with `ExtendedCacheInterface` + `ExtendedTagAwareCacheInterface` for pools implementing both intrefaces: the proposed ones + PSR-6's `CacheItemPoolInterface`.

File: src/Symfony/Contracts/Cache/ExtendedTagAwareCacheInterface.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Contracts\Cache;
+
+use Psr\Cache\CacheItemPoolInterface;
+
+/**
+ * An interface that extends TawAwareCacheInterface with the methods from PSR-6's CacheItemPoolInterface.
```
Comment (Cydonia7) [inline src/Symfony/Contracts/Cache/ExtendedTagAwareCacheInterface.php]: There's a typo: `Taw` => `Tag`, also present in the PR description.

Comment (nicolas-grekas) [reply src/Symfony/Contracts/Cache/ExtendedTagAwareCacheInterface.php]: fixed thanks

File: src/Symfony/Contracts/Cache/TagAwareCacheInterface.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+interface TagAwareCacheInterface extends CacheInterface
+{
+    /**
+     * Invalidates cached items using tags.
+     *
+     * When implemented on a PSR-6 pool, invalidation should not apply
+     * to deferred items. Instead, they should be committed as usual.
+     * This allows replacing old tagged values by news ones without
```
[the remark this answers is not in the document]
Comment (nicolas-grekas) [reply src/Symfony/Contracts/Cache/TagAwareCacheInterface.php]: fixed, thanks

File: src/Symfony/Contracts/Cache/ExtendedTagAwareCacheInterface.php [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+namespace Symfony\Contracts\Cache;
+
+use Psr\Cache\CacheItemPoolInterface;
+
+/**
+ * An interface that extends TagAwareCacheInterface with the methods from PSR-6's CacheItemPoolInterface.
+ */
+interface ExtendedTagAwareCacheInterface extends TagAwareCacheInterface, CacheItemPoolInterface
```
Comment (Nyholm) [inline src/Symfony/Contracts/Cache/ExtendedTagAwareCacheInterface.php]: As discussed: Remove this.

File: src/Symfony/Contracts/Cache/ExtendedCacheInterface.php [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+namespace Symfony\Contracts\Cache;
+
+use Psr\Cache\CacheItemPoolInterface;
+
+/**
+ * An interface that extends CacheInterface with the methods from PSR-6's CacheItemPoolInterface.
+ */
+interface ExtendedCacheInterface extends CacheInterface, CacheItemPoolInterface
```
Comment (Nyholm) [inline src/Symfony/Contracts/Cache/ExtendedCacheInterface.php]: We should remove this.

**Review by Nyholm — APPROVED** [review]
Great. 
I had a discussion with @nicolas-grekas. 
Approved with changes

**Comment by nicolas-grekas** [comment]
Thanks @Nyholm , PR updated.

**Comment by Nyholm** [comment]
Thank you. Im 👍

**Comment by nicolas-grekas** [comment]
(rebased)

**Comment by nicolas-grekas** [comment]
Addressed, thanks.

**Review by fabpot — APPROVED** [review]

**Pull request merged**